### ISSUE TITLE
Oops on ubuntu 22.04

### ISSUE THREAD

**Issue by theundefined** [cschramm/irda#4]
laptop T430 upgraded from ubuntu 18.04 to 22.04. On 18.04 irda works fine, hardware didn't changed.
Now i compiled irda as dkms. After plug a stick - it load module correcty, but when i try to run `irattach irda0 -s` there is Oops:

```
[  141.560533] usb 1-1.2: new low-speed USB device number 6 using ehci-pci
[  146.779521] usb 1-1.2: New USB device found, idVendor=07d0, idProduct=4100, bcdDevice= 0.00
[  146.779527] usb 1-1.2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[  146.779529] usb 1-1.2: Product: USB to IRDA
[  146.779530] usb 1-1.2: Manufacturer: Kingsun Co, Ltd.
[  152.494860] NET: Registered PF_IRDA protocol family
[  152.498078] KingSun/Dazzle IRDA/USB found at address 6, Vendor: 7d0, Product: 4100
[  152.502974] net irda0: IrDA: Registered KingSun/Dazzle device irda0
[  152.511697] usbcore: registered new interface driver ksdazzle-sir
[  167.028411] BUG: kernel NULL pointer dereference, address: 000000000000049c
[  167.028418] #PF: supervisor read access in kernel mode
[  167.028421] #PF: error_code(0x0000) - not-present page
[  167.028424] PGD 0 P4D 0
[  167.028428] Oops: 0000 [#1] SMP PTI
[  167.028433] CPU: 1 PID: 6649 Comm: irattach Tainted: G         C OE     5.15.0-27-generic #28-Ubuntu
[  167.028437] Hardware name: LENOVO 2344BMU/2344BMU, BIOS G1ETB4WW (2.74 ) 09/25/2017
[  167.028440] RIP: 0010:ip_mc_up+0x44/0xc0
[  167.028449] Code: 04 24 b9 c0 0c 00 00 31 d2 be e0 00 00 01 4c 89 e7 48 8b 80 30 05 00 00 49 c7 44 24 50 12 7a 00 00 49 c7 44 24 58 c4 09 00 00 <8b> 80 9c 04 00 00 41 88 44 24 60 e8 dc e5 ff ff 49 8b 5c 24 18 48
[  167.028453] RSP: 0018:ffffb814033bfbe0 EFLAGS: 00010246
[  167.028457] RAX: 0000000000000000 RBX: 0000000000000001 RCX: 0000000000000cc0
[  167.028460] RDX: 0000000000000000 RSI: 00000000010000e0 RDI: ffff90f04c0b3c00
[  167.028463] RBP: ffffb814033bfbf0 R08: 0000000000000000 R09: 0000000000000001
[  167.028465] R10: ffffb814033bfc50 R11: 0000000000000000 R12: ffff90f04c0b3c00
[  167.028468] R13: ffff90f04a86d000 R14: 00000000fffffff6 R15: ffffb814033bfde0
[  167.028471] FS:  00007fa62aaf0740(0000) GS:ffff90f36e640000(0000) knlGS:0000000000000000
[  167.028475] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  167.028478] CR2: 000000000000049c CR3: 00000001be79c001 CR4: 00000000001706e0
[  167.028481] Call Trace:
[  167.028483]  <TASK>
[  167.028487]  inetdev_event+0x342/0x530
[  167.028494]  raw_notifier_call_chain+0x49/0x60
[  167.028500]  ? devinet_init_net+0x2c0/0x2c0
[  167.028505]  ? raw_notifier_call_chain+0x49/0x60
[  167.028510]  call_netdevice_notifiers_info+0x52/0xa0
[  167.028515]  __dev_notify_flags+0x58/0xe0
[  167.028520]  dev_change_flags+0x51/0x60
[  167.028524]  dev_ifsioc+0x305/0x530
[  167.028528]  ? netdev_name_node_lookup_rcu+0x6b/0x80
[  167.028533]  ? __cond_resched+0x1a/0x50
[  167.028539]  dev_ioctl+0x161/0x550
[  167.028544]  sock_do_ioctl+0xa7/0x100
[  167.028548]  sock_ioctl+0xef/0x310
[  167.028552]  ? __audit_syscall_entry+0xde/0x120
[  167.028557]  __x64_sys_ioctl+0x91/0xc0
[  167.028563]  do_syscall_64+0x5c/0xc0
[  167.028568]  ? syscall_exit_to_user_mode+0x27/0x50
[  167.028571]  ? do_syscall_64+0x69/0xc0
[  167.028575]  entry_SYSCALL_64_after_hwframe+0x44/0xae
[  167.028579] RIP: 0033:0x7fa62ac0daff
[  167.028583] Code: 00 48 89 44 24 18 31 c0 48 8d 44 24 60 c7 04 24 10 00 00 00 48 89 44 24 08 48 8d 44 24 20 48 89 44 24 10 b8 10 00 00 00 0f 05 <41> 89 c0 3d 00 f0 ff ff 77 1f 48 8b 44 24 18 64 48 2b 04 25 28 00
[  167.028586] RSP: 002b:00007ffecbb40f00 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[  167.028591] RAX: ffffffffffffffda RBX: 00000000000000c1 RCX: 00007fa62ac0daff
[  167.028593] RDX: 00007ffecbb40f60 RSI: 0000000000008914 RDI: 0000000000000001
[  167.028596] RBP: 0000000000000001 R08: 000000000000000a R09: 0000000000000000
[  167.028599] R10: 0000000000004000 R11: 0000000000000246 R12: 00007ffecbb40f60
[  167.028601] R13: 00000000ffffffff R14: 000055727a204048 R15: ffffffffffffffff
[  167.028606]  </TASK>
[  167.028608] Modules linked in: ksdazzle_sir(OE) irda(OE) iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi rfcomm nf_conntrack_netlink br_netfilter xt_CHECKSUM xt_MASQUERADE nft_chain_nat nf_nat bridge stp llc xfrm_user xfrm_algo vmw_vsock_vmci_transport vsock l2tp_ppp vmw_vmci l2tp_netlink l2tp_core ip6_udp_tunnel udp_tunnel pppox ccm overlay toshiba_acpi industrialio sparse_keymap toshiba_haps acpi_call(OE) nfnetlink_log hp_accel lis3lv02d cmac algif_hash algif_skcipher af_alg bnep ip6t_REJECT nf_reject_ipv6 xt_hl ip6_tables ip6t_rt ipt_REJECT nf_reject_ipv4 xt_LOG nf_log_syslog nft_limit xt_limit xt_addrtype xt_tcpudp xt_conntrack nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 nft_compat nft_counter nf_tables nfnetlink binfmt_misc dm_thin_pool btusb btrtl uvcvideo btbcm btintel videobuf2_vmalloc bluetooth videobuf2_memops videobuf2_v4l2 videobuf2_common ecdh_generic videodev ecc mc xfs dm_cache_smq dm_cache dm_persistent_data dm_bio_prison dm_bufio intel_rapl_msr mei_hdcp
[  167.028688]  intel_rapl_common x86_pkg_temp_thermal intel_powerclamp coretemp snd_hda_codec_hdmi kvm_intel kvm rapl intel_cstate think_lmi firmware_attributes_class input_leds wmi_bmof serio_raw snd_ctl_led at24 snd_hda_codec_realtek snd_hda_codec_generic snd_hda_intel snd_intel_dspcfg thinkpad_acpi nvram ledtrig_audio snd_intel_sdw_acpi snd_seq_midi iwldvm snd_hda_codec snd_seq_midi_event mac80211 snd_hda_core snd_rawmidi libarc4 snd_hwdep mei_me snd_pcm snd_seq iwlwifi snd_seq_device snd_timer cfg80211 snd mei soundcore platform_profile mac_hid sch_fq_codel cuse ipmi_devintf ipmi_msghandler msr parport_pc ppdev lp parport binder_linux ashmem_linux(C) ip_tables x_tables autofs4 dm_raid btrfs blake2b_generic zstd_compress dm_crypt raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid0 multipath linear raid1 i915 nouveau mxm_wmi crct10dif_pclmul crc32_pclmul i2c_algo_bit drm_ttm_helper ttm ghash_clmulni_intel drm_kms_helper aesni_intel
[  167.028742]  i2c_i801 crypto_simd cryptd psmouse i2c_smbus syscopyarea sysfillrect sdhci_pci sysimgblt ahci fb_sys_fops cec libahci rc_core lpc_ich drm cqhci sdhci e1000e xhci_pci xhci_pci_renesas wmi video
[  167.028758] CR2: 000000000000049c
[  167.028760] ---[ end trace b7025948ed3bc007 ]---
[  167.028762] RIP: 0010:ip_mc_up+0x44/0xc0
[  167.028766] Code: 04 24 b9 c0 0c 00 00 31 d2 be e0 00 00 01 4c 89 e7 48 8b 80 30 05 00 00 49 c7 44 24 50 12 7a 00 00 49 c7 44 24 58 c4 09 00 00 <8b> 80 9c 04 00 00 41 88 44 24 60 e8 dc e5 ff ff 49 8b 5c 24 18 48
[  167.028768] RSP: 0018:ffffb814033bfbe0 EFLAGS: 00010246
[  167.028770] RAX: 0000000000000000 RBX: 0000000000000001 RCX: 0000000000000cc0
[  167.028771] RDX: 0000000000000000 RSI: 00000000010000e0 RDI: ffff90f04c0b3c00
[  167.028773] RBP: ffffb814033bfbf0 R08: 0000000000000000 R09: 0000000000000001
[  167.028774] R10: ffffb814033bfc50 R11: 0000000000000000 R12: ffff90f04c0b3c00
[  167.028775] R13: ffff90f04a86d000 R14: 00000000fffffff6 R15: ffffb814033bfde0
[  167.028777] FS:  00007fa62aaf0740(0000) GS:ffff90f36e640000(0000) knlGS:0000000000000000
[  167.028779] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  167.028780] CR2: 000000000000049c CR3: 00000001be79c001 CR4: 00000000001706e0
[  232.528641] L1TF CPU bug present and SMT on, data leak possible. See CVE-2018-3646 and https://www.kernel.org/doc/html/latest/admin-guide/hw-vuln/l1tf.html for details.
[  233.409383] usb 1-1.2: ksdazzle_rcv_irq: urb asynchronously failed - -71
[  233.541653] usb 1-1.2: USB disconnect, device number 6
[  234.578635] usb 1-1.2: ksdazzle_hard_xmit: failed tx_urb submit: -19

```
unplug donge doesn't help.

**Comment by theundefined** [comment]
The same problem with oops i found on second dongle, so i think it's more related to ubuntu kernel (5.15). Is anyone using any usb dongle with that kernel?

**Comment by cschramm** [comment]
I can confirm the issue with an `irtty_sir` on 5.17, so it seems rather generic. :confused: I actually did not actively use an IrDa dongle since ~5.4. Will try to investigate.

**Comment by cschramm** [comment]
34cf097d419515f3819da186a3d47c0f97814358 is at fault, I guess the `ip_ptr` use.

**Comment by cschramm** [comment]
Yep, just setting `ip_ptr` as in https://github.com/cschramm/irda/commit/34cf097d419515f3819da186a3d47c0f97814358#diff-aa8f59ee60bfc8cadc98e510a5660ece5661b5dec640dcb1543e1082c3ee860bR131 (even without reading it) is enough to cause the problem. I guess I'll revert back to `atalk_ptr`. That means that `CONFIG_ATALK` is required in 5.17+ but I do not see another easy solution right now.

**Comment by cschramm** [comment]
Fixed in `master` and set the `linux-4.15` tag to a proper revision.
